core/src/main/scala/kafka/log/LogSegment.scala

```diff
@@ -303,15 +311,25 @@ class LogSegment(val log: FileRecords,
     // We may need to reload the max timestamp after truncation.
     if (maxTimestampSoFar >= 0)
       loadLargestTimestamp(readToLogEnd = true)
+    _nextOffset = math.max(baseOffset, math.min(_nextOffset, offset))
     bytesTruncated
   }
 
   /**
-   * Calculate the offset that would be used for the next message to be append to this segment.
+   * The offset that would be used for the next message to be appended to this segment
+   */
+  def nextOffset(): Long = {
+    if (_nextOffset == INVALID_NEXT_OFFSET)
+      _nextOffset = nextOffsetFromLog()
```

Comment:
After thinking about this more, I agree that it is safer to keep less state in the class unless `nextOffset()` is called frequently. Thus I agree the suggested approach is simpler. Thanks @junrao.